readme.md
# MRML

[![Crates.io](https://img.shields.io/crates/d/mrml)](https://crates.io/crates/mrml)
![Crates.io](https://img.shields.io/crates/v/mrml)
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fjdrouet%2Fmrml.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Fjdrouet%2Fmrml?ref=badge_shield)

[![.github/workflows/main.yml](https://github.com/jdrouet/mrml/actions/workflows/mrml-core-main.yml/badge.svg)](https://github.com/jdrouet/mrml/actions/workflows/mrml-core-main.yml)
[![codecov](https://codecov.io/gh/jdrouet/mrml/branch/main/graph/badge.svg?token=SIOPR0YWZA)](https://codecov.io/gh/jdrouet/mrml)

[![Average time to resolve an issue](http://isitmaintained.com/badge/resolution/jdrouet/mrml.svg)](http://isitmaintained.com/project/jdrouet/mrml "Average time to resolve an issue")
[![Percentage of issues still open](http://isitmaintained.com/badge/open/jdrouet/mrml.svg)](http://isitmaintained.com/project/jdrouet/mrml "Percentage of issues still open")
[![Maintainability](https://api.codeclimate.com/v1/badges/7ed23ef670d076ab69a4/maintainability)](https://codeclimate.com/github/jdrouet/mrml/maintainability)

## Introduction

This project is a reimplementation of the nice [MJML markup language](https://documentation.mjml.io/) in Rust.

## How to use it in my code

Update your `cargo.toml`:

```toml
[dependencies]
mrml = "1.2"
serde = { version = "1.0", features = ["derive"] }
```

Create your `main.rs`:

```rust
use mrml;

fn main() {
    let root = mrml::parse("<mjml><mj-body></mj-body></mjml>").expect("parse template");
    let opts = mrml::prelude::render::Options::default();
    match root.render(&opts) {
        Ok(content) => println!("{}", content),
        Err(_) => println!("couldn't render mjml template"),
    };
}
```

Available options are:

| Name                 | Comment                                             | Default value                                                                                        |
| -------------------- | --------------------------------------------------- | ---------------------------------------------------------------------------------------------------- |
| `disable_comments`   | Strip comments out of rendered HTML                 | `false`                                                                                              |
| `social_icon_origin` | Custom URL for fetching social icons                | `None`                                                                                               |
| `fonts`              | Default fonts imported in the HTML rendered by MJML | [See default options](https://github.com/jolimail/mrml-core/blob/main/src/prelude/render.rs#L33-L54) |

## Why?

- A Node.js server rendering an MJML template takes around 20 MB of RAM at startup and 130 MB under stress test. In Rust, less than 1.7 MB at startup and a bit less that 3 MB under stress test. The Rust version can also handle twice as many requests per second. You can perform the benchmarks by running `bash script/run-bench.sh`.
- The JavaScript implementation cannot be run in the browser; the Rust one (and WebAssembly one) can be.

## You want to contribute?

Feel free to read our [contributing](./contributing.md) section and the [code of conduct](./code-of-conduct.md).

## Performance

With the same Linux amd64 machine, to render the amario template

- Node: 606.59ms
- Rust: 3.48ms

## Missing implementations

- `mj-style[inline]`: not yet implemented. It requires parsing the generated html to apply the inline styles afterward (that's how it's done in mjml) which would kill the performances. Applying it at render time would improve the performance but it would still require to parse the CSS.
- `mj-include`: not yet implemented. It requires to handle loading remote templates when using mrml in a wasm (browser or server side) format, which implies being able to load from a different location (`file://`, `https://`, relative, etc).

## Who is using MRML?

[<img src="https://www.blizzstatic.com/www/marketing/images/logo.svg" height="22px" />](https://www.blizzfull.com/)

<i>If you are using MRML and want to be added to this list, don't hesitate to create an issue or open a pull request.</i>

## What is using MRML?

[mjml_nif](https://github.com/adoptoposs/mjml_nif) - Elixir library

[mrml-ruby](https://github.com/hardpixel/mrml-ruby) - Ruby library

[mjml-python](https://github.com/mgd020/mjml-python) - Python library

<i>If you are using MRML and want to be added to this list, don't hesitate to create an issue or open a pull request.</i>

## You want to sponsor us?

[<img src="https://liberapay.com/assets/liberapay/icon-v2_white-on-yellow.svg?etag=.Z1LYSBJ8Z6GWUeLUUEf2XA~~" height="35px" />](https://liberapay.com/jdrouet/)
[<img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Buy Me A Coffee" height="35px" />](https://www.buymeacoffee.com/jdrouet)

<i>Thanks to [zachzurn](https://github.com/zachzurn).</i>

## License

[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fjdrouet%2Fmrml.svg?type=large)](https://app.fossa.com/projects/git%2Bgithub.com%2Fjdrouet%2Fmrml?ref=badge_large)


changelog.md
# Changelog

All notable changes to this project will be documented in this file.

## [2.0.0-rc4] - 2023-08-17

### Bug Fixes

- Disable_comments is in camelCase
- Script elements should remain with open and close element (#320)

### Documentation

- Update links in readme
- Update code in example

### Features

- Bump mrml version
- Remove randomness when generating component ids (#308)
- Create wasm package (#312)
- Add mj-include feature (#316)
- Create multi include loader (#322)

### Miscellaneous Tasks

- Allow mastodon notification to fail
- Implement real tool for comparing html (#107)
- Move license files
- Update cargo.toml to publish
- Update cargo.toml to publish
- Bump to 2.0.0-rc4

### Refactor

- Put cows in tag implementation 🐮  (#111)
- Prepare for monotrepo
- Move back mrml-cli to repository
- Format code using rustfmt and create config
- Apply clippy suggestions
- Use a single MrmlParser structure with a visitor pattern (#317)

### Styling

- Apply clippy suggestions

### Testing

- Add node example with a test (#313)
- Add browser example (#315)
- Update node example to have test command
- Add test with disabling comments

### Build

- Bump quote from 1.0.28 to 1.0.29 (#101)
- Bump serde from 1.0.164 to 1.0.166 (#102)
- Bump thiserror from 1.0.40 to 1.0.41 (#103)
- Bump serde_json from 1.0.99 to 1.0.100 (#104)
- Bump colored from 2.0.0 to 2.0.4 (#110)
- Bump serde from 1.0.166 to 1.0.167 (#108)
- Bump serde from 1.0.167 to 1.0.168 (#112)
- Bump thiserror from 1.0.41 to 1.0.43 (#109)
- Bump proc-macro2 from 1.0.63 to 1.0.64 (#113)
- Bump serde_json from 1.0.100 to 1.0.102 (#114)
- Bump semver from 5.7.1 to 5.7.2 in /scripts (#115)
- Update dependabot configuration
- Update env_logger requirement in /packages/mrml-cli (#309)
- Bump thiserror from 1.0.43 to 1.0.44 (#305)
- Update clap to 4.3
- Update indexmap to 2.0

### Ci

- Move code-checking out of every workflow (#314)
- Report mrml-cli metric to codebench (#321)

### Doc

- Update readme

## [2.0.0-rc3] - 2023-06-27

### Features

- Add fonts option (#90)

### Build

- Bump serde from 1.0.158 to 1.0.159 (#69)
- Bump indexmap from 1.9.2 to 1.9.3 (#68)
- Bump mockito from 1.0.0 to 1.0.1 (#66)
- Bump proc-macro2 from 1.0.53 to 1.0.54 (#67)
- Bump serde_json from 1.0.94 to 1.0.95 (#71)
- Bump reqwest from 0.11.15 to 0.11.16 (#70)
- Bump mockito from 1.0.1 to 1.0.2 (#72)
- Bump proc-macro2 from 1.0.54 to 1.0.56 (#73)
- Bump serde from 1.0.159 to 1.0.160 (#74)
- Bump serde_json from 1.0.95 to 1.0.96 (#75)
- Bump h2 from 0.3.15 to 0.3.17 (#76)
- Bump reqwest from 0.11.16 to 0.11.17 (#77)
- Bump serde from 1.0.160 to 1.0.162 (#78)
- Bump quote from 1.0.26 to 1.0.27 (#79)
- Bump serde from 1.0.162 to 1.0.163 (#80)
- Bump proc-macro2 from 1.0.56 to 1.0.57 (#81)
- Bump reqwest from 0.11.17 to 0.11.18 (#82)
- Bump proc-macro2 from 1.0.57 to 1.0.58 (#83)
- Bump quote from 1.0.27 to 1.0.28 (#86)
- Bump proc-macro2 from 1.0.58 to 1.0.59 (#85)
- Bump criterion from 0.4.0 to 0.5.1 (#87)
- Bump url from 2.3.1 to 2.4.0 (#89)
- Bump serde from 1.0.163 to 1.0.164 (#91)
- Bump proc-macro2 from 1.0.59 to 1.0.60 (#92)
- Bump ureq from 2.6.2 to 2.7.0 (#93)
- Bump serde_json from 1.0.96 to 1.0.97 (#94)
- Bump mockito from 1.0.2 to 1.1.0 (#95)
- Bump ureq from 2.7.0 to 2.7.1 (#96)
- Bump proc-macro2 from 1.0.60 to 1.0.63 (#99)
- Bump serde_json from 1.0.97 to 1.0.99 (#98)

## [2.0.0-rc2] - 2023-03-25

### Bug Fixes

- Replace deprecated methods

### Documentation

- Add link to mjml

### Features

- Mark IncludeLoaderError cayse as Send and Sync (#52)

### Miscellaneous Tasks

- Enable the orderedmap feature by default

### Build

- Bump mockito from 0.32.0 to 0.32.1 (#40)
- Bump mockito from 0.32.1 to 0.32.2 (#41)
- Bump mockito from 0.32.2 to 0.32.3 (#42)
- Bump syn from 1.0.107 to 1.0.108 (#43)
- Bump syn from 1.0.108 to 1.0.109 (#44)
- Bump mockito from 0.32.3 to 0.32.4 (#46)
- Bump thiserror from 1.0.38 to 1.0.39 (#48)
- Bump serde_json from 1.0.93 to 1.0.94 (#47)
- Bump serde from 1.0.152 to 1.0.153 (#49)
- Bump serde from 1.0.153 to 1.0.154 (#50)
- Bump darling from 0.14.3 to 0.14.4 (#51)
- Bump quote from 1.0.23 to 1.0.25 (#54)
- Bump serde from 1.0.154 to 1.0.155 (#55)
- Bump mockito from 0.32.4 to 1.0.0 (#53)
- Bump quote from 1.0.25 to 1.0.26 (#58)
- Bump serde from 1.0.155 to 1.0.156 (#57)
- Bump thiserror from 1.0.39 to 1.0.40 (#60)
- Bump serde from 1.0.156 to 1.0.157 (#61)
- Bump reqwest from 0.11.14 to 0.11.15 (#62)
- Bump serde from 1.0.157 to 1.0.158 (#63)
- Bump proc-macro2 from 1.0.52 to 1.0.53 (#65)

## [2.0.0-rc1] - 2023-02-12

### Documentation

- Add link to python library

### Features

- First implementation of mj-include (#23)
- Create an include-loader to load from the filesystem (#32)
- Create an http-loader (#34)

### Miscellaneous Tasks

- Publish on mastodon after releasing a version
- Replace print boilerplate with a procedural macro (#10)
- Check features with cargo hack (#12)
- Create a macro to generate From and as_element (#13)
- Replace parse boilerplate with a proc macro (#11)
- Macro cleanup (#14)
- Implement a proc macro to handle json conversion (#18)
- Replace some existing code with macro for parsing (#20)
- Rename components using camel case (#21)
- Camelcase  (#22)
- Create a struct to simplify code (#24)
- Split in multiple modules (#25)
- Improve code coverage (#27)
- Update semantic commit checker (#33)
- Remove feature powerset for code checking
- Implement stderr for errors (#38)
- Rename and set versions
- Update cargo.toml files with required fields

### Build

- Bump proc-macro2 from 1.0.50 to 1.0.51 (#30)
- Bump serde_json from 1.0.91 to 1.0.92 (#29)
- Bump darling from 0.14.2 to 0.14.3 (#31)
- Add deps scope to semantic configuration
- Bump serde_json from 1.0.92 to 1.0.93 (#35)
- Bump mockito from 0.31 to 0.32

## [1.2.11] - 2023-01-11

### Bug Fixes

- Run and push coverage
- Upgrade typescript from 3.7.5 to 3.9.7

src/lib.rs
//! This project is a reimplementation of the nice [MJML](https://mjml.io/) markup language in Rust.
//!
//! [![codecov](https://codecov.io/gh/jolimail/mrml-core/branch/main/graph/badge.svg?token=SIOPR0YWZA)](https://codecov.io/gh/jolimail/mrml-core)
//! [![.github/workflows/main.yml](https://github.com/jolimail/mrml-core/actions/workflows/main.yml/badge.svg)](https://github.com/jolimail/mrml-core/actions/workflows/main.yml)
//! [![Average time to resolve an issue](http://isitmaintained.com/badge/resolution/jolimail/mrml-core.svg)](http://isitmaintained.com/project/jolimail/mrml-core "Average time to resolve an issue")
//! [![Percentage of issues still open](http://isitmaintained.com/badge/open/jolimail/mrml-core.svg)](http://isitmaintained.com/project/jolimail/mrml-core "Percentage of issues still open")
//! [![Maintainability](https://api.codeclimate.com/v1/badges/7ed23ef670d076ab69a4/maintainability)](https://codeclimate.com/github/jolimail/mrml-core/maintainability)
//!
//! To use it you can simply update your `Cargo.toml` by adding
//! ```toml
//! [dependencies]
//! mrml = "1.2"
//! serde = { version = "1.0", features = ["derive"] }
//! ```
//!
//! And you can then just create a `main.rs` with the following code
//! ```rust
//! let root = mrml::parse("<mjml><mj-body></mj-body></mjml>").expect("parse template");
//! let opts = mrml::prelude::render::Options::default();
//! match root.render(&opts) {
//!     Ok(content) => println!("{}", content),
//!     Err(_) => println!("couldn't render mjml template"),
//! };
//! ```
//!
//! You can also use the `mj-include` component by specifying a [loader](crate::prelude::parse).
//! ```rust
//! use mrml::prelude::parse::ParserOptions;
//! use mrml::prelude::parse::memory_loader::MemoryIncludeLoader;
//! use std::rc::Rc;
//!
//! let loader = MemoryIncludeLoader::from(vec![("partial.mjml", "<mj-button>Hello</mj-button>")]);
//! let options = Rc::new(ParserOptions {
//!     include_loader: Box::new(loader),
//! });
//! match mrml::parse_with_options("<mjml><mj-head /><mj-body><mj-include path=\"partial.mjml\" /></mj-body></mjml>", options) {
//!     Ok(_) => println!("Success!"),
//!     Err(err) => eprintln!("Something went wrong: {err:?}"),
//! }
//! ```
//!
//! ### Why?
//!
//! A Node.js server rendering an MJML template takes around **20 MB** of RAM at startup and **130 MB** under stress test.
//! In Rust, less than **1.7 MB** at startup and a bit less that **3 MB** under stress test.
//! The Rust version can also handle twice as many requests per second.
//! You can perform the benchmarks by running `bash script/run-bench.sh`.
//!
//! Also, the JavaScript implementation cannot be run in the browser; the Rust one (and WebAssembly one) can be.

pub mod comment;
pub mod mj_accordion;
pub mod mj_accordion_element;
pub mod mj_accordion_text;
pub mod mj_accordion_title;
pub mod mj_attributes;
pub mod mj_attributes_all;
pub mod mj_attributes_class;
pub mod mj_attributes_element;
pub mod mj_body;
pub mod mj_breakpoint;
pub mod mj_button;
pub mod mj_carousel;
pub mod mj_carousel_image;
pub mod mj_column;
pub mod mj_divider;
pub mod mj_font;
pub mod mj_group;
pub mod mj_head;
pub mod mj_hero;
pub mod mj_image;
pub mod mj_include;
pub mod mj_navbar;
pub mod mj_navbar_link;
pub mod mj_preview;
pub mod mj_raw;
pub mod mj_section;
pub mod mj_social;
pub mod mj_social_element;
pub mod mj_spacer;
pub mod mj_style;
pub mod mj_table;
pub mod mj_text;
pub mod mj_title;
pub mod mj_wrapper;
pub mod mjml;
pub mod node;
pub mod prelude;
pub mod text;

mod helper;
mod macros;

#[cfg(feature = "parse")]
/// Function to parse a raw mjml template with some parsing [options](crate::prelude::parse::ParserOptions).
/// This function is just an alias to [the `Mjml::parse_with_options` function](crate::mjml::Mjml).
///
/// You can specify the kind of loader mrml needs to use for loading the content of
/// [`mj-include`](crate::mj_include) elements.
///
/// You can take a look at the available loaders [here](crate::prelude::parse).
///
/// ```rust
/// use mrml::prelude::parse::ParserOptions;
/// use mrml::prelude::parse::memory_loader::MemoryIncludeLoader;
/// use std::rc::Rc;
///
/// let options = Rc::new(ParserOptions {
///     include_loader: Box::new(MemoryIncludeLoader::default()),
/// });
/// match mrml::parse_with_options("<mjml><mj-head /><mj-body /></mjml>", options) {
///     Ok(_) => println!("Success!"),
///     Err(err) => eprintln!("Something went wrong: {err:?}"),
/// }
/// ```
pub fn parse_with_options<T: AsRef<str>>(
    input: T,
    opts: std::rc::Rc<crate::prelude::parse::ParserOptions>,
) -> Result<mjml::Mjml, prelude::parse::Error> {
    mjml::Mjml::parse_with_options(input, opts)
}

#[cfg(feature = "parse")]
/// Function to parse a raw mjml template using the default parsing [options](crate::prelude::parse::ParserOptions).
///
/// ```rust
/// match mrml::parse("<mjml><mj-head /><mj-body /></mjml>") {
///     Ok(_) => println!("Success!"),
///     Err(err) => eprintln!("Something went wrong: {err:?}"),
/// }
/// ```
pub fn parse<T: AsRef<str>>(input: T) -> Result<mjml::Mjml, prelude::parse::Error> {
    mjml::Mjml::parse(input)
}
